File: TP3/TP3/src/Ej7/BankAccount.java
package Ej7;

public abstract class BankAccount {
    private final int id;
    protected double balance;

    public void deposit(double amount){
        balance += amount;
    }
    protected abstract boolean canExtract(double amount);

    public double getBalance(){
        return balance;
    }

    public void extract(double amount){
        if(canExtract(amount)){
            balance -= amount;
        }
    }
    protected BankAccount(int id){
        this.id = id;
        balance = 0;
    }

    
    @Override
    public String toString(){
        return "Cuenta %d con saldo %.2f".formatted(id,balance);
    }
}

File: TP8/EJ5/bank_account.rb
# frozen_string_literal: true

# Abstract Class
class BankAccount
  def initialize
    raise 'Cannot initialize abstract class BankAccount'
  end

  def create(id)
    @id = id
    @balance = 0
  end

  attr_reader :balance

  def extract?; end

  def amount_valid (amount)
    (amount.is_a? Numeric) || amount.positive?
  end

  def deposit(amount)
    return unless amount_valid amount

    @balance += amount
  end

  def extract(amount)
    raise 'no funds' unless (amount_valid amount) && extract?

    @balance -= amount
  end

  #Cuenta 1001 con saldo 100.0
  def to_s
    "Cuenta #{@id} con saldo #{@balance}"
  end
  private :create, :amount_valid, :extract?
end
